### DISCUSSION TITLE
[Idea] Accept non-regression fixes and new hooks in a PrestaShop patch releases

### DISCUSSION THREAD

**Discussion by kpodemski** [PrestaShop/PrestaShop#32206] [Ideas]
### The problem

I would like to discuss with all of you the idea of being more flexible with patch versions of PrestaShop. The idea is to accept non-regression fixes and new hooks in PrestaShop patch releases.

### The solution

Currently, patch versions are about fixing regressions introduced in the same minor version of PrestaShop and accepting critical or security issues.

This leads us to a situation where, very often, impactful bugfix submitted by the community is merged into the following major or minor version. It takes months or even years to have it available in the release.

A possible solution would be to expand the scope of accepted fixes beyond just regression issues, and include bug fixes that can significantly enhance the stability of the software and overall user experience. How to do that? We could evaluate whether the submitted fix fits right into the patch version. If there is too much risk of the fix introducing another regression, it would be scheduled for the next major/minor release.

Second subject. Let’s start accepting new hooks in patch versions.

I know. New hooks sound like a new feature. It does not follow SemVer. The reality is that developers often need those to fix issues in their modules, and to stop using overrides that are considered a primary pain point in the stability of the software. Disadvantages? We would need to release `autoupgrade` module more often, to follow the changes between patch releases. How we could do that? We could use the same strategy as we now have in the `classic-theme` repository. We could always have a stable branch ready to be released with only those changes that are required for upcoming releases.

I’m open to your feedback and suggestions.

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by FabienPapet** [comment]
I agree with the first part, accepting bug fixes in a patch is the way to go IMO. But new hooks are basically new features, and do not belong to patch version. Let's follow standards for once, please.

100% for accepting non-regressions.
100% against adding new hooks.

**Comment by jolelievre** [comment]
I think both are a good idea, adding new hook shouldn't impact the stability of the software as long as they are not used anyway. So the core as standalone shouldn't have problem with new hooks. However, but it's also valid on version under development, we still need to asses all the time if the hook is legitimate and if it's behaviour could impact the stability in a bad way (ex: add a hook in the middle of cart rule processing is probably a bad idea in patch, minor and major versions).

Also if we decide to accept SOME bug fixes, but not all of them (which also sounds legit if we still want to favor the stability of our patch versions) we should have clear and objective criteria to decide if a bug can be accepted or not. This will prevent frustration and infinite debate.

**Comment by matks** [comment]
> we should have clear and objective criteria to decide if a bug can be accepted or not

I think it should be quite easy to explain: users want patch versions to be as reliable as possible. When I upgrade from PrestaShop 8.0.1 to 8.0.2 I want to feel safe that I'm just strenghtening the software, not taking a risk. So any code change that goes into a patch version should, in theory
- be well designed and tested
- do not impact an area / component / workflow / feature too big (example a huge modification of the HTML layout structure), rather focus on surgical improvements targeting a well defined zone
- do not introduce what we believe to be a risky change

The question that must be asked for each PR submitted against a patch version should be: "from what I know, if we merge this PR, are we making the software better (= we deliver more value than we remove), do I think it is more reliable and stable than before?"

**Reply by jolelievre** [reply]
We all know that the opinion of committers/maintainers is regularly questioned, I just want to anticipate the frustration of `Why do you refuse my PR you accepted this one? It works perfectly on my shop there is no risk at all! Do your job and merge ok!` 😅

Which is why the `we believe to be a risky change` can be quite subjective, but I know having clear criteria is quite complicated since each PR is a special case potentially.

**Reply by eternoendless** [reply]
At the end of the day people will have to trust and accept maintainers' decisions.

**Comment by eternoendless** [comment]
Let me provide a little context. Historically, the reasoning behind only accepting regressions in patch versions was:

* An extremely high level of tech debt, coupled with low test coverage and limited knowledge of legacy code by a relatively new team, meant that "quick" bug fixes often brought new regressions (especially in 1.6.1.x)
* A high amount of PRs in patch versions diverted our efforts away from minor versions, where the largest improvements are made

We need to keep the risks in mind when making our choice. I think the software and the processes have been improved enough to allow for low-risk fixes to be accepted in patch versions, whether they fix regressions or old bugs. That said, we've already delayed 8.1 because of patch releases. We need to carefully choose our battles so as to avoid making things worse.

Strictly speaking, new hooks are indeed features. However, time-to-market is also an important factor that we have to seriously keep in mind. As long as we won't be able to release minor versions every month or so, we need to find an alternative. **The one thing to have in mind about this is that hooks might have a non-trivial impact on performance, especially if triggered during a loop.**

Regarding the upgrade module: it's not a tragedy if we miss adding the hook in an upgrade script – if a hook doesn't exist, they are automatically added into the hooks table when used. Adding them into the table is only important when it's a display hook, so that people can use the BO feature to hook their module in it, and get a description.

**Comment by Hlavtox** [comment]
### Bugfixes
I think that we could be able to distinguish the danger of the bugfix and make sure we touch only small-scoped things.

### Hooks
I think that new hooks are perfectly safe and we don't even need to maintain multiple autoupgrade branches. Other new features in patch versions - no.

**Comment by eternoendless** [comment]
After a discussion I had today with @kpodemski, my suggestion on bug fixes:

* Major and critical bug fixes go in the patch version (regressions or not) - unless they are judged as "too risky"
* All other bug fixes go by default in the develop branch - unless maintainers think it's appropriate to "promote" them to another branch.
* New hooks can be accepted in patch branches as long as they aren't risky. Merging a new hook in a patch branch shouldn't necessarily mean that a patch version will be released though.

All of the above obviously concerns the actively maintained branch only, and not the extended support branch (which remains limited to critical fixes only).

**Comment by kpodemski** [comment]
@PrestaShop/prestashop-maintainers @PrestaShop/scoped-maintainers 

What do you think about @eternoendless suggestion? Do you think we agree on this?

@jolelievre in the guides, this is exactly what I would like to wrote:

> Major and critical bug fixes go in the patch version (regressions or not) - unless they are judged as "too risky"
> All other bug fixes go by default in the develop branch - unless maintainers think it's appropriate to "promote" them to another branch.
> New hooks can be accepted in patch branches as long as they aren't risky. Merging a new hook in a patch branch shouldn't necessarily mean that a patch version will be released though.

I don't think we can come up with set of rules on what makes bug fixes acceptable on the patch branch. We will do that on a case by case basis.

**Reply by MatShir** [reply]
Even if it's not a written rule, there's a correlation between major and technical risk when you look at the remaining major bugs. 
So we won't have much PR in the patch, more than usual since the critical (regressions or not has already been accepted).
So if a minor bug isn't risky, why refuse it? But some bugs can be really annoying, even if they don't have a major impact on the merchant's business.

**Comment by ghost** [comment]
It's a good idea, but I think we need rules
<< We will do that on a case by case basis. >> 
With all the new contributors there are many rules that are no longer respected. 
For example an issue per PR, some time ago even for a wording mistake an issue was needed.
The hooks were not accepted either, which could frustrate contributors, which is why rules seem obvious to me.

**Reply by kpodemski** [reply]
> For example an issue per PR, some time ago even for a wording mistake an issue was needed.

I think you will agree that this rule was a bit off, and there really was no need for that, we had to be a bit more pragmatic. I think the same goes for this proposal, otherwise, we will end up having a bunch of rules where it is impossible to put every single scenario. Too generic rules = more issues. I don't know, maybe we will come up with some idea for this one.

**Reply by matks** [reply]
We have tried before to write rules that work all the time, for all contributions, for all usecases. We have failed 😅 .

To be able to write a rule and then follow it 100% we would need to be able to anticipate _every_ possibility and write what to do in that case. IMO PrestaShop is too big, too complex, and changes too fast for that.

So the only thing we can do, I believe, is trust people 😉 . When PR (...) will be submitted, maintainers will do the right choice (accept or reject), based on their experience and context. But we cannot write a rule now and expect people to follow it 100% in 6 months when faced with a situation they did not expect and we did not anticipate

**Comment by kpodemski** [comment]
It's been a while. The rule to accept some non-regression fixes and new hooks to patch versions has been in place for some time already, and it did help quite a few times! We'll update information in the developer documentation and on the project's website to reflect what's been decided.

We don't have guidelines about what bug fixes won't be accepted, we don't have something like that for hooks, we're flexible and open to discussion here. As @matks mentioned, it's impossible to anticipate every existing scenario and find an answer to it.

I will close this discussion. If anyone wants to discuss this topic further, please [join us on Slack](https://www.prestashop-project.org/slack/) and send me a message :)
